vector/src/main/java/im/vector/VectorApp.java

```diff
@@ -279,8 +288,83 @@ public void onCreate() {
+
+                try {
+                    // Get all activities from app manifest
+                    if (null == mActivityInfos) {
+                        mActivityInfos = packageManager.getPackageInfo(getPackageName(), PackageManager.GET_ACTIVITIES).activities;
+                    }
+
+                    // Get all running activities on app task
+                    // and compare to activities declared in manifest
+                    ActivityManager manager = (ActivityManager) getSystemService(ACTIVITY_SERVICE);
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+                        List<ActivityManager.AppTask> tasks = manager.getAppTasks();
+
+                        for (ActivityManager.AppTask task : tasks) {
+                            ComponentName topActivity = task.getTaskInfo().topActivity;
+
+                            if (null != topActivity && isPotentialMaliciousActivity(topActivity, mActivityInfos)) {
+                                return true;
+                            }
+                        }
+                    } else {
+                        List<ActivityManager.RunningTaskInfo> runningTaskInfos = manager.getRunningTasks(10);
+
+                        for (ActivityManager.RunningTaskInfo runningTaskInfo : runningTaskInfos) {
+                            ComponentName topActivity = runningTaskInfo.topActivity;
+
+                            if (null != topActivity && isPotentialMaliciousActivity(topActivity, mActivityInfos)) {
+                                return true;
+                            }
+                        }
+                    }
+                } catch (PackageManager.NameNotFoundException e) {
+                    Log.e(LOG_TAG, "Error " + e.getMessage(), e);
+                }
+
+                return false;
+            }
+
+            /**
+             * Detect potential malicious activity.
+             * Check if the activity running with package name task affinity is declared in app manifest.
+             *
+             * @param activity      the activity of the task
+             * @param activityInfos the list of activities declared in app manifest
+             * @return true if the activity is potentially malicious
+             */
+            private boolean isPotentialMaliciousActivity(ComponentName activity, ActivityInfo[] activityInfos) throws PackageManager.NameNotFoundException {
+                String taskAffinity = getPackageManager().getActivityInfo(activity, 0).taskAffinity;
+
+                if (taskAffinity != null && taskAffinity.equals(getPackageName())) {
+                    for (ActivityInfo activityInfo : activityInfos) {
+                        if (activityInfo.name.equals(activity.getClassName())) {
+                            return false;
+                        }
+                    }
+                } else {
+                    return false;
+                }
+                return true;
```

Comment:
Perhaps it would be better to define a boolean value `isPotentialMaliciousActivity` initialized with true

```
                boolean isPotentialMaliciousActivity = true;
                String taskAffinity = getPackageManager().getActivityInfo(activity, 0).taskAffinity;
                if (taskAffinity != null) {
                    // Check whether the task affinity matches with the package name
                    if (taskAffinity.equals(getPackageName())) {
                        // Check whether the activity is legitimate (declared in the manifest)
                        for (ActivityInfo activityInfo : activityInfos) {
                            if (activityInfo.name.equals(activity.getClassName())) {
                                isPotentialMaliciousActivity = false;
                                break;
                            }
                        }
                    } else {
                        // The activity is considered safe when its task affinity doesn't correspond to the package name
                        isPotentialMaliciousActivity = false;
                    }
                }
                return isPotentialMaliciousActivity;
```

Reply:
isPotentialMaliciousActivity variable was added.